core/network/impl/loopback_stream.cpp

```diff
@@ -0,0 +1,184 @@
+  using E = kagome::network::LoopbackStream::Error;
+  switch (e) {
+    case E::INVALID_ARGUMENT:
+      return "invalid argument was passed";
+    case E::IS_READING:
+      return "read was called before the last read completed";
+    case E::IS_CLOSED_FOR_READS:
+      return "this stream is closed for reads";
+    case E::IS_WRITING:
+      return "write was called before the last write completed";
+    case E::IS_CLOSED_FOR_WRITES:
+      return "this stream is closed for writes";
+    case E::IS_RESET:
+      return "this stream was reset";
+    case E::INTERNAL_ERROR:
+      return "internal error happened";
+  }
+  return "unknown error";
+}
+
+namespace kagome::network {
+  LoopbackStream::LoopbackStream(libp2p::peer::PeerInfo own_peer_info)
+      : own_peer_info_(std::move(own_peer_info)) {}
+
+  bool LoopbackStream::isClosedForRead() const {
+    return !is_readable_;
+  };
+
+  bool LoopbackStream::isClosedForWrite() const {
+    return !is_writable_;
+  };
+
+  bool LoopbackStream::isClosed() const {
+    return isClosedForRead() && isClosedForWrite();
+  };
+
+  void LoopbackStream::close(VoidResultHandlerFunc cb) {
+    is_writable_ = false;
+    cb(outcome::success());
+  };
+
+  void LoopbackStream::reset() {
+    is_reset_ = true;
+  };
+
+  void LoopbackStream::adjustWindowSize(uint32_t new_size,
+                                        VoidResultHandlerFunc cb){};
+
+  outcome::result<bool> LoopbackStream::isInitiator() const {
+    return outcome::success(false);
+  };
+
+  outcome::result<libp2p::peer::PeerId> LoopbackStream::remotePeerId() const {
+    return outcome::success(own_peer_info_.id);
+  };
+
+  outcome::result<libp2p::multi::Multiaddress> LoopbackStream::localMultiaddr()
+      const {
+    return outcome::success(own_peer_info_.addresses.front());
+  };
+
+  outcome::result<libp2p::multi::Multiaddress> LoopbackStream::remoteMultiaddr()
+      const {
+    return outcome::success(own_peer_info_.addresses.front());
+  }
+
+  void LoopbackStream::read(gsl::span<uint8_t> out,
+                            size_t bytes,
+                            libp2p::basic::Reader::ReadCallbackFunc cb) {
+    read(out, bytes, std::move(cb), false);
+  }
+
+  void LoopbackStream::readSome(gsl::span<uint8_t> out,
+                                size_t bytes,
+                                libp2p::basic::Reader::ReadCallbackFunc cb) {
+    read(out, bytes, std::move(cb), true);
+  }
+
+  void LoopbackStream::write(gsl::span<const uint8_t> in,
+                             size_t bytes,
+                             libp2p::basic::Writer::WriteCallbackFunc cb) {
+    if (is_reset_) {
+      return cb(Error::IS_RESET);
+    }
+    if (!is_writable_) {
+      return cb(Error::IS_CLOSED_FOR_WRITES);
+    }
+    if (bytes == 0 || in.empty() || static_cast<size_t>(in.size()) < bytes) {
+      return cb(Error::INVALID_ARGUMENT);
+    }
+    if (is_writing_) {
+      return cb(Error::IS_WRITING);
+    }
```

Comment:
We can remove is_writing flag and corresponding error I think. I cannot imagine situation when we can end up in this if as is_writing flag will be true only during execution of this function. It only might be helpful if write invoked from different threads, but then such flag is not thread safe anyway

Reply:
All this method should do is write to buffer_ and invoke WriteCallBackFunc